src/iamclient/iamclient.cpp

```diff
@@ -5,331 +5,604 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <Poco/Pipe.h>
+#include <Poco/PipeStream.h>
+#include <Poco/Process.h>
+#include <Poco/StreamCopier.h>
+
+#include <utils/exception.hpp>
 #include <utils/grpchelper.hpp>
 
 #include "iamclient.hpp"
 #include "log.hpp"
 
 /***********************************************************************************************************************
- * Public
+ * Static
  **********************************************************************************************************************/
 
-aos::Error IAMClient::Init(const Config& config, aos::iam::certhandler::CertHandlerItf& certHandler,
-    aos::cryptoutils::CertLoaderItf& certLoader, aos::crypto::x509::ProviderItf& cryptoProvider, bool provisioningMode)
+static iamanager::v5::CPUInfo ConvertCpuInfoToProto(const aos::CPUInfo& val)
+{
+    iamanager::v5::CPUInfo result;
+
+    result.set_model_name(val.mModelName.CStr());
+    result.set_num_cores(val.mNumCores);
+    result.set_num_threads(val.mNumThreads);
+    result.set_arch(val.mArch.CStr());
+    result.set_arch_family(val.mArchFamily.CStr());
+
+    return result;
+}
+
+static iamanager::v5::PartitionInfo ConvertPartitionInfoToProto(const aos::PartitionInfo& val)
+{
+    iamanager::v5::PartitionInfo result;
+
+    result.set_name(val.mName.CStr());
+
+    for (const auto& type : val.mTypes) {
+        result.add_types(type.CStr());
+    }
+
+    result.set_total_size(val.mTotalSize);
+
+    return result;
+}
+
+static iamanager::v5::NodeAttribute ConvertNodeAttributeToProto(const aos::NodeAttribute& val)
+{
+    iamanager::v5::NodeAttribute result;
+
+    result.set_name(val.mName.CStr());
+    result.set_value(val.mValue.CStr());
+
+    return result;
+}
+
+static iamanager::v5::NodeInfo ConvertNodeInfoToProto(const aos::NodeInfo& val)
 {
+    iamanager::v5::NodeInfo result;
 
+    result.set_id(val.mID.CStr());
+    result.set_type(val.mType.CStr());
+    result.set_name(val.mName.CStr());
+    result.set_status(val.mStatus.ToString().CStr());
+    result.set_os_type(val.mOSType.CStr());
+
+    for (const auto& item : val.mCPUs) {
+        *result.add_cpus() = ConvertCpuInfoToProto(item);
+    }
+
+    result.set_max_dmips(val.mMaxDMIPS);
+    result.set_total_ram(val.mTotalRAM);
+
+    for (const auto& item : val.mPartitions) {
+        *result.add_partitions() = ConvertPartitionInfoToProto(item);
+    }
+
+    for (const auto& item : val.mAttrs) {
+        *result.add_attrs() = ConvertNodeAttributeToProto(item);
+    }
+
+    return result;
+}
+
+static common::v1::ErrorInfo ConvertAosErrorToProto(aos::Error error)
+{
+    common::v1::ErrorInfo result;
+
+    result.set_aos_code(static_cast<int32_t>(error.Value()));
+    result.set_exit_code(error.Errno());
+
+    if (!error.IsNone()) {
+        aos::StaticString<aos::cMaxErrorStrLen> message;
```

Comment:
message is empty.
I suggest something like this:

`        auto err = message.Convert(error);

        result.set_message(err.IsNone() ? message.CStr() : error.Message());`

Reply:
fixed